src/eos-metrics-instrumentation.c

```diff
@@ -39,17 +39,32 @@
 #define BOOT_COUNT_KEY "boot_count"
 
 /*
- * Started when a user logs in and stopped when that user logs out.
- * Payload contains the user ID of the user that logged in.
- * (Thus the payload is a GVariant containing a single unsigned 32-bit integer.)
+ * The event ID to record user session's alive time from login to logout.
+ * https://azafea.readthedocs.io/en/latest/events.html#azafea.event_processors.endless.metrics.v3.model.DailySessionTime
  */
-#define USER_IS_LOGGED_IN "add052be-7b2a-4959-81a5-a7f45062ee98"
+#define DAILY_SESSION_TIME "5dc0b53c-93f9-4df0-ad6f-bd25e9fe638f"
 
 #define MIN_HUMAN_USER_ID 1000
 
 static EinsPersistentTally *persistent_tally;
 
-static GData *humanity_by_session_id;
+typedef struct _UserSession {
+  guint32 user_id;
+  EmtrAggregateTimer *aggregate_timer;
+  gatomicrefcount rc;
+} UserSession;
+
+/*
+ * An user only has a signle UserSession. Have a hashtable for the login users.
+ * The new login seesion will increase the UserSession's reference count rc.
+ */
+static GHashTable *humanity_by_session_id;
```

Comment:
What are the keys and values of this hash table?

It seems that the keys are the session IDs (though obfuscated in a way I don't quite understand – if they are strings then the keys should be strings and the hash and equality functions should be g_str_hash and g_str_equal) and the values are:

- uint32 user_id
- timer
- refcount

But I think these UserSession values are shared between all keys with the same user ID.

I would structure this differently. The hashtable keys should be the user ID itself (stored with GUINT_TO_POINTER (uid), and key function `g_direct_hash` and value function `g_direct_equal`). The values should be a structure :

```
typedef struct _UserSessions {
  EmtrAggregateTimer *timer;
  // no need to be atomic, this is a single-threaded app
  int n_sessions
} UserSessions;
```

When a new session is added:

- get its user ID
- look the user ID up in the hash table
  - If present, add 1 to the counter
  - If not present, start a timer, and insert it inside a new UserSessions struct into the hash table, with `GUINT_TO_POINTER (user_id)` as the key

When a session is removed:

- get its user ID
- look it up in the hash table to get the UserSessions struct
- subtract 1 from its `n_sessions`
- if it is now  0, stop the timer and remove the uid from the hash table

Reply:
Change the hash table's key to user ID in the new commit.